presto-main/src/main/java/com/facebook/presto/cost/StatsAndCosts.java

```diff
@@ -77,15 +78,15 @@ public StatsAndCosts getForSubplan(PlanNode root)
         return new StatsAndCosts(filteredStats.build(), filteredCosts.build());
     }
 
-    public static StatsAndCosts create(PlanNode root, StatsProvider statsProvider, CostProvider costProvider)
+    public static StatsAndCosts create(PlanNode root, StatsProvider statsProvider, CostProvider costProvider, Session session, boolean useCache)
     {
         Iterable<PlanNode> planIterator = Traverser.forTree(PlanNode::getSources)
                 .depthFirstPreOrder(root);
         ImmutableMap.Builder<PlanNodeId, PlanNodeStatsEstimate> stats = ImmutableMap.builder();
         ImmutableMap.Builder<PlanNodeId, PlanCostEstimate> costs = ImmutableMap.builder();
         for (PlanNode node : planIterator) {
-            stats.put(node.getId(), statsProvider.getStats(node));
-            costs.put(node.getId(), costProvider.getCost(node));
+            stats.put(node.getId(), useCache ? session.getPlanNodeEstimateMap().getOrDefault(node.getId(), PlanNodeStatsEstimate.unknown()) : statsProvider.getStats(node));
```

Comment:
why not calculate them fresh if the stats aren't there like a loading cache would.  two benefits:
1. we won't need the on and off code paths just to disable the feature for tests.  
2. would prevent confusion if a node happened to get a new id after any cost-based optimizations ran, but it's otherwise something we would know the stats for.

Reply:
> why not calculate them fresh if the stats aren't there like a loading cache would. two benefits:

The reason is because the calculation of stats will invalidate the recorded stats of child nodes.
Take the example in the description:
```
presto:tpch> explain (type distributed) select name, price from (select custkey, sum(totalprice) price from orders group by custkey) t join customer using(custkey);
                                                                                                                                                   >
--------------------------------------------------------------------------------------------------------------------------------------------------->
 Fragment 0 [SINGLE]                                                                                                                               >
     Output layout: [name, sum]                                                                                                                    >
     Output partitioning: SINGLE []                                                                                                                >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                 >
     - Output[PlanNodeId 17][name, price] => [name:varchar(25), sum:double]                                                                        >
             price := sum (1:41)                                                                                                                   >
         - RemoteSource[1] => [name:varchar(25), sum:double]                                                                                       >
                                                                                                                                                   >
 Fragment 1 [HASH]                                                                                                                                 >
     Output layout: [name, sum]                                                                                                                    >
     Output partitioning: SINGLE []                                                                                                                >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                 >
     - InnerJoin[PlanNodeId 366][("custkey_9" = "custkey")][$hashvalue, $hashvalue_30] => [name:varchar(25), sum:double]                           >
             Estimates: {source: CostBasedSourceInfo, rows: 1,001 (31.29kB), cpu: 1,309,503.16, memory: 35,640.00, network: 335,820.00}            >
             Distribution: PARTITIONED                                                                                                             >
         - RemoteSource[2] => [custkey_9:bigint, name:varchar(25), $hashvalue:bigint]                                                              >
         - Project[PlanNodeId 489][projectLocality = LOCAL] => [custkey:bigint, sum:double, $hashvalue_30:bigint]                                  >
                 $hashvalue_30 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(custkey), BIGINT'0')) (1:143)                               >
             - Aggregate(FINAL)[custkey][PlanNodeId 3] => [custkey:bigint, sum:double]                                                             >
                     Estimates: {source: CostBasedSourceInfo, rows: 990 (61.88kB), cpu: ?, memory: ?, network: ?}                                  >
                     sum := "presto.default.sum"((sum_25)) (1:69)                                                                                  >
                 - LocalExchange[PlanNodeId 446][HASH][$hashvalue_27] (custkey) => [custkey:bigint, sum_25:double, $hashvalue_27:bigint]           >
                     - RemoteSource[3] => [custkey:bigint, sum_25:double, $hashvalue_28:bigint]                                                    >
                                                                                                                                                   >
 Fragment 2 [SOURCE]                                                                                                                               >
     Output layout: [custkey_9, name, $hashvalue_26]                                                                                               >
     Output partitioning: HASH [custkey_9][$hashvalue_26]                                                                                          >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                 >
     - ScanProject[PlanNodeId 8,487][table = TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=customer,>
             Estimates: {source: CostBasedSourceInfo, rows: 1,500 (60.06kB), cpu: ?, memory: ?, network: ?}/{source: CostBasedSourceInfo, rows: ? (>
             $hashvalue_26 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(custkey_9), BIGINT'0')) (1:128)                                 >
             LAYOUT: tpch.customer{}                                                                                                               >
             custkey_9 := custkey:bigint:0:REGULAR (1:128)                                                                                         >
             name := name:varchar(25):1:REGULAR (1:128)                                                                                            >
                                                                                                                                                   >
 Fragment 3 [SOURCE]                                                                                                                               >
     Output layout: [custkey, sum_25, $hashvalue_29]                                                                                               >
     Output partitioning: HASH [custkey][$hashvalue_29]                                                                                            >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                 >
     - Project[PlanNodeId 488][projectLocality = LOCAL] => [custkey:bigint, sum_25:double, $hashvalue_29:bigint]                                   >
             $hashvalue_29 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(custkey), BIGINT'0')) (1:112)                                   >
         - Aggregate(PARTIAL)[custkey][PlanNodeId 450] => [custkey:bigint, sum_25:double]                                                          >
                 sum_25 := "presto.default.sum"((totalprice)) (1:69)                                                                               >
             - TableScan[PlanNodeId 0][TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=orders, analyze>
                     Estimates: {source: CostBasedSourceInfo, rows: 15,000 (395.51kB), cpu: ?, memory: ?, network: ?}                              >
                     LAYOUT: tpch.orders{}                                                                                                         >
                     custkey := custkey:bigint:1:REGULAR (1:96)                                                                                    >
                     totalprice := totalprice:double:3:REGULAR (1:96)                                                                              >
                                                                                                                                                   >
                                                                                                                                                   >
(1 row)
```
At the end, when we estimate the stats of the output node, it will try to estimate the stats of the join node, which then go to the aggregation node. CBO return empty stats for the final aggregation node, which will be recorded and invalidate the stats previously recorded for both join and aggregation nodes. Because we are currently recording the latest estimation stats in our cache.

I think one solution is to not always record the latest estimation, but to skip if the new stats is unknown. However, my concern is that this may skip some legit changes, for example if this happens not in the end but during optimizations.

Currently I found the biggest and only usage of the printed estimation stats is to help me to debug optimization behaviour, so I think keeping the stats used during optimization is more important. Having the stats provider to calculate the stats in the end, which is not the stats used during query optimization can be misleading and not so useful to me.